### PR TITLE
[FEATURE] - Add support for temporary credentials

### PR DESCRIPTION
**Pull request by JKCT**
Hello!
Been checking out Gleam and it's very promising - congratulations on the stable release!

I use AWS at `$DAYJOB` and was looking to try using Gleam with it.  It appears that there is not yet an [AWS SDK](https://aws.amazon.com/developer/tools/) for Gleam, and making AWS API requests directly is non-trivial because each request must be signed with [AWS Signature V4](https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html) which is time consuming to implement from scratch.
My hope is to contribute a new AWS SDK package for Gleam which will hopefully be beneficial to the community.

I was delighted to see there was prior art for Gleam with this package, and it was extremely useful in modeling some of the best practices that may have been tricky for a first time contributor ie. dependency injection.

The current plan is to make a new package (`aws_sdk_gleam` or similar) that depends on this one for request signing.  
Before doing that I wanted to ensure this package supports [temporary credentials (session token)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html) in addition to the access keys it already supports.  

Temporary session credentials are the primary type of credentials used with AWS and are preferred for security as they expire where access keys do not.  Common uses for AWS temporary session credentials:
- When assuming an IAM Role
- When authenticating from inside an AWS Service ie. in EC2, ECS, Lambda session credentials are automatically configured as environment variables

Open to feedback and suggestions! Please let me know if there is any guidance on changes that should be made to this pull request and I will implement them.

# New Features
- Now supports [temporary credentials (session token)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html).
- Added method to support `Request(String)` as well as the existing `Request(BitArray)`

## Other Changes
- Updated README
- Updated tests to cover all branch combinations
- Added documentation comments for module/functions
- Refactored some code in `src/aws4_request.gleam` for readability

## Version Revision
Major - [Breaking changes to public API](https://semver.org/)

Previously this package had a single public API in `sign` which accepted only `Request(BitArray)`.  Based on the [gleam_httpc package](https://hexdocs.pm/gleam_httpc/gleam/httpc.html#send) appears that `Request(String)` is also a common type convention which could be supported.

`gleam_httpc ` distinguishes between the two public APIs by having:
- `send`: `Request(String)`
- `send_bits`: `Request(BitArray)`

Unfortunately here the convention is reversed in the main branch where `sign` is `Request(BitArray)`, so it cannot be matching without breaking changes to the public API.  This update has the following public APIs:
- `sign_string`: `Request(String)`  (`sign` would also work here)
- `sign_bits`: `Request(BitArray)`

## Testing
These methods have been tested locally with both access keys and temporary credentials.  
Unit tests have been updated to provide coverage without committing secret values.

# Open Questions
It appears there is not yet a `time` package in the [standard library](https://hexdocs.pm/gleam_stdlib/) which seems like it would be a good addition.  The interface enabled by the standard library's [uri package](https://hexdocs.pm/gleam_stdlib/gleam/uri.html) provided an awesome user experience here where `Request(BitArray)` can be signed by another module and then easily sent with a choice of servers, not just httpc.

It seems like it would be beneficial to define a standard datetime interface in the standard library to enable a universal type definition.  Besides a timestamp being used here for signing the request, several AWS API endpoints accept/send datetime type fields.

Unrelated - I see there is a `src/aws4_request_ffi.erl` file which doesn't seem to be in use for FFI, perhaps it can be removed?

# Next Steps
Once this package supports temporary session credentials, I plan to make a code generator that creates fully typed JSON encoders/decoders programmatically using the [JSON documentation for boto3](https://github.com/boto/botocore/blob/develop/botocore/data/cloudcontrol/2021-09-30/service-2.json).  Boto3 is the AWS SDK for python and the SDK that powers [AWS CLI](https://aws.amazon.com/cli/).  It appears to have the most complete "documentation" via these JSON files, more so than the official docs or other SDK repositories.

Generating all the service code programmatically should enable:
- All 200+ AWS services to be supported quickly
- New services to be added easily by re-running the generator

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,5 +52,3 @@ pub fn main() {
   httpc.send(signed_request)
 }
 ```
-
-Further documentation can be found at <https://hexdocs.pm/buckets>.
```
Comment (JKCT) [inline README.md:49]: It appears this link is inactive - maybe there is an updated URL to use?

File: src/aws4_request.gleam [hunk 2]
```diff
@@ -1,28 +1,116 @@
...
+///
+///   Parameters:
+/// request: The request to sign of type Request(BitArray).
+/// date_time: The current datetime in #((year, month, day), (hour, minute, second)) format.
+/// access_key_id: The AWS credential access key ID being used to authenticate the request.
+/// secret_access_key: The AWS credential secret access key being used to authenticate the
+/// request.
+/// session_token: (Optional) The security token value if using temporary credentials from AWS STS.
+/// region: The AWS region that the request will be made to.
+/// service: The AWS service name that the request is for.
```
Comment (JKCT) [inline src/aws4_request.gleam:29]: I saw that parameters needed to be documented but I was unable to find an example of the correct format in the language documentation or in source code for the libraries I've seen so far.

Originally I put these `///` comments directly above the line of each parameter, but `gleam format` puts them all up here so I went with this, but please link me if there is an example of how to properly do this!

**Review by lpil — COMMENTED** [review]
Hello! Thank you for this.

This PR does lots of things, we'll need to have one change or feature per PR for review. The leftover Erlang code and the outdated link I've updated on main 👍 

Could you explain to me how one gets a session token and when I would want to use one? Thank you.

P.S. Very excited to hear about AWS bindings being made! I think it would be nice to have a package per-service to avoid having folks pull in huge amounts of unused code when they only want to use a couple services.

**Comment by JKCT** [comment]
Thank you for the quick response and the updates to main!

> we'll need to have one change or feature per PR for review
Good call, these changes started smaller and spiraled as these things tend to do, apologies for that.

I'm closing this PR and opened a new one that is just for adding support for `Request(String)` . 
I'm planning to make at least two more PRs after, one for commenting all the code and one for adding session token support. I can decompose the changes further if needed just let me know.

> Could you explain to me how one gets a session token and when I would want to use one? 
Great question, this is the key to any changes so it should be explained and understood first.

The [AWS Official Documentation for Identity and Access Management (IAM)](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html) is authoritative but very extensive, I'll do my best to summarize the relevant points here and in the future PRs.

There are three different types of IAM "Identities" that can be used for authentication/access to AWS, root user, IAM Users and IAM Roles.  AWS strongly discourages use of the root user because it's permissions cannot be limited, and instead recommends creating scoped Users and Roles with unique credentials.  IAM Users use permanent access keys (made up of an `Access Key ID` and `Secret Access Key` parts). IAM Roles use temporary session tokens (made up of an `Access Key ID` and `Secret Access Key` and `Session Token`) that expire in 12 hours or less (configurable). 

While personal users are likely to be familiar with access keys due to ease of use, professional users usually and sometimes exclusively use temporary sessions.  This is because temporary sessions are an AWS security best practice as they expire and certain actions can _only _be done via IAM Roles.
Source: https://docs.aws.amazon.com/IAM/latest/UserGuide/id.html

Here's a few common examples of things that can only be done with IAM Roles and session tokens from the link above:

1. `Federated user access`:  I have a CLI tool written in gleam I want to use to place files in an S3 Bucket.  I authenticate to my company's AWS account using our Single Sign On (SSO) provider using my Active Directory username and password which then provides me with session token credentials for an IAM Role.   I may not have access to access keys at all in this scenario.
2. `Cross-account access`: I have automated my CLI tool written in gleam that places files in buckets.  It now runs in our GitHub Actions CICD flow.  During this flow, it will copy files from our CICD AWS account to our separate Development environment AWS account.  It does this by preforming [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) which only provides temporary session credentials.
3. `Applications running on Amazon EC2`: I have created a server written in gleam that saves files to an S3 Bucket.  I launch this application on EC2 (compute platform).  Each EC2 instance has an IAM Role.  I grant the EC2 Role access to the S3 bucket.  The application running on EC2 is provided temporary session credentials via environment variables that AWS sets on the instance.

Let me know if there are any parts I can explain in further detail.


> I think it would be nice to have a package per-service
This is excellent feedback, I'll experiment in this direction to see if there are any potential tradeoffs to this approach.

**Pull request closed without merging**